src/components/workflow/workflow.service.ts

```diff
@@ -0,0 +1,866 @@
+    try{
+      await this.db
+        .query()
+        .match([
+          [
+            ...matchSession(session),
+            relation('in', '', 'member'),
+            node('sg', 'SecurityGroup'),
+            relation('in', '', 'securityGroup', {
+              active: true
+            }),
+            node('fromState', 'State', {
+              id: input.fromStateId
+            }),
+            relation('out', 'rel', 'nextPossibleState', {
+              active: true
+            }),
+            node('toState', 'State', {
+              id: input.toStateId
+            })
+          ],
+        ])
+        .detachDelete('rel')
+        .run();
+
+    } catch (e) {
+      this.logger.warning('failed to remove possible state from state', {
+        exception: e
+      });
+      throw e;
+    }
+  }
+
+  // there will be more than one required field relationship between a state node and a base node.
+  // this is so each required field can be queried without inspecting the property name in app code.
+  // addRequiredFieldToState
+  async addField(session: ISession, input: RequiredField): Promise<void>{
+    try{
+      const field = await this.db
+        .query()
+        .match([
+          [
+            node('token', 'Token', {
+              active: true,
+              value: session.token
+            }),
+            relation('in', '', 'token', {
+              active: true
+            }),
+            node('user'),
+            relation('in', '', 'admin', {
+              active: true
+            }),
+            node('baseNode'),
+            relation('out', '', `${input.propertyName}`),
+            node('property')
+          ]
+        ])
+        .return({
+          property: 'property'
+        })
+        .first();
+
+      if ( !field ) {
+        throw new NotFoundException('could not find such field existing.');
+      }
+
+      await this.db
+        .query()
+        .match([
+          [
+            node('token', 'Token', {
+              active: true,
+              value: session.token
+            }),
+            relation('in', '', 'token', {
+              active: true
+            }),
+            node('user'),
+            relation('in', '', 'admin', {
+              active: true
+            }),
+            node('baseNode')
+          ],
+          [
+            node('state', 'State', {
+              id: input.stateId
+            })
+          ]
+        ])
+        .merge([
+          node('baseNode'),
+          relation('in', '', 'requiredProperty', {
+            value: input.propertyName
+          }),
+          node('state')
+        ])
+        .first();
+
+    } catch (e) {
+      this.logger.warning('could not add field to state', {
+        exception: e
+      });
+      throw e;
+    }
+  }
+
+  // listAllRequiredFieldsInAState
+  async listFields(session: ISession, stateId: string): Promise<RequiredFieldListOutput>{
```

Comment:
The frontend needs this info to render the form correctly so let's make it available to all users who can access the fields. no need to check for any permission except that they have read access to each field. That will be security groups and permission nodes.

Reply:
also should change the name to listRequiredFields.